File: DataStructures/Stack/Stack.py
# collections.deque로 스택 구현
from collections import deque


class Stack:
    def __init__(self):
        self.stack = deque()

    def __str__(self):
        return "".join(str(self.stack))

    # 스택이 비었으면 True 반환, 그렇지 않으면 False 반환
    def isEmpty(self):
        return False if len(self.stack) else True

    # 스택의 끝에 data 삽입
    def push(self, data):
        self.stack.append(data)

    # 스택에 값이 있다면 끝에서 data를 꺼내고 반환
    def pop(self):
        if self.isEmpty():
            return "Stack is Empty"
        else:
            return self.stack.pop()

    # 스택에 값이 있다면 끝의 data를 반환
    def peek(self):
        if self.isEmpty():
            return "Stack is Empty"
        else:
            return self.stack[-1]

    # 스택의 크기 반환
    def size(self):
        return len(self.stack)


if __name__ == '__main__':
    myStack = Stack()
    myStack.push(1)
    myStack.push(2)
    myStack.push(3)
    print("myStack:", myStack)
    print("myStack의 맨 끝:", myStack.peek())
    print("myStack에서 pop한 값:", myStack.pop())
    print("myStack에서 pop한 값:", myStack.pop())
    print("myStack에서 pop한 값:", myStack.pop())
    print("myStack에서 pop한 값:", myStack.pop())

File: DataStructures/Stack/Stack.cpp
#include <iostream>

// 스택의 최대 크기를 1000으로 설정
#define MAX 1000

using namespace std;

class Stack
{
    public:
        Stack()
        {
            mTop = -1; 
        }
        bool Push(int); // 스택의 가장 맨 위에 값을 추가하는 함수
        int Pop();  // 스택의 가장 맨 위 요소를 삭제하고 반환하는 함수
        int Peek(); // 스택의 가장 맨 위 요소를 삭제없이 반환하는 함수
        bool IsEmpty(); // 스택이 비었으면 true, 그렇지 않으면 false를 반환하는 함수

    private:
        int mStack[MAX];
        int mTop;   // 스택의 가장 맨 위의 요소의 인덱스

};

bool Stack::Push(const int data)
{
    if (mTop >= MAX)
    {
        cout << "Stack is Full";
        return false;
    }
    else
    {
        mStack[++mTop] = data;
        return true;
    }
}

int Stack::Pop()
{
    if (mTop < 0)
    {
        cout << "Stack is Empty" << '\n';
        return false;
    }
    else
    {
        return mStack[mTop--];
    }
}

int Stack::Peek()
{
    if (mTop < 0)
    {
        cout << "Stack is Empty" << '\n';
        return false;
    }
    else
    {
        return mStack[mTop];
    }
}

bool Stack::IsEmpty()
{
    return (mTop < 0);
}


int main()
{
    Stack mystack;
    mystack.Push(1);   
    mystack.Push(2);    
    mystack.Push(3);
    cout << mystack.Peek() << '\n';
    cout << mystack.Pop() << '\n';
    cout << mystack.Pop() << '\n';
    cout << mystack.Pop() << '\n';
    cout << mystack.Pop() << '\n';
    cout << mystack.IsEmpty() << '\n';

    return 0;
}
